[export] fix node.users when inlining HOOs

pianpwk opened the pull request:
The process of inlining HOO subgraphs (e.g. set_grad_enabled) seems to break node.users when a node is present in multiple subgraphs, for example:
```
class SetGradCase(torch.nn.Module):
    def forward(self, x):
        _x = x.shape[0] + 2
        _xx = _x + 2
        with torch.no_grad():
            y = _x * 4
        return _xx, y
```

The `_x` node contains 2 users (_xx and y) after being inlined, but with inspection it seems to only contain y as a user.

Previously we were completely clearing node.users for output nodes in HOO subgraphs before inlining them - we should just be deleting the subgraph output nodes

larryliu0820 commented on test/export/test_export.py:
```diff
@@ -799,6 +799,44 @@ def forward(self, x, weight, bias):
                 actual_result.append(node.meta.get("torch_fn"))
         self.assertEqual(actual_result, expected_result)
 
+    def test_hoo_inline_users_issue(self):
```
test_hop?

pianpwk replied:
higher order op? I have no idea what acronym is the right one hahaha

mcr229 commented:
why are there 23 commits lol

pianpwk commented:
> why are there 23 commits lol

very very good question lol

ydwu4 commented on torch/_export/passes/replace_with_hop_pass_util.py:
```diff
@@ -171,6 +171,15 @@ def _replace_with_hop_pass_helper(
             )
             setattr(new_gm, node.target, new_subgm)
 
+    # The process of inlining nodes seems to break node.users when a node is present in
```
Instead of fixing it up here, can we try to figure out why the inline call doesn't properly change the users and fix it up there? Then we have a better inline function that's less buggy, which improves the composability of the system.

pianpwk commented on torch/_export/utils.py:
```diff
@@ -410,10 +410,9 @@ def node_inline_(call_mod_node: torch.fx.Node) -> None:
             elif isinstance(new_output, (list, tuple)):
-                # Clear the users of the output node and set
-                # the users to be the users of original call_module node.
+                # Pop subgraph output node from users.
```
Shouldn't completely clear this, just pop the subgraph output node

ydwu4 reviewed (approved):
Looks great! Thanks for taking the extra step and figure out the true root cause.

pianpwk committed:
[export] fix node.users when inlining HOOs (#133144)

Summary:
The process of inlining HOO subgraphs (e.g. set_grad_enabled) seems to break node.users when a node is present in multiple subgraphs, for example:
```
class SetGradCase(torch.nn.Module):
    def forward(self, x):
        _x = x.shape[0] + 2
        _xx = _x + 2
        with torch.no_grad():
            y = _x * 4
        return _xx, y
```

The _x node is present in both a `grad_enabled=True` and `grad_enabled=False` subgraph, with different users (_xx and y, respectively). This breaks when we inline them back, and is a big problem if we later DCE, DCE-ing nodes that very clearly have users.

This patches up node.users at the end of `replace_with_hop_pass_helper`.


Differential Revision: D61053873
